### ISSUE TITLE
Xcode Metal Capture?

### ISSUE THREAD

**Issue by MattGuerrette** [KhronosGroup/MoltenVK#301]
I can't for the life of me get the Xcode frame capture to work with MoltenVK. I've tried the LunarG source code as well as another popular online Vulkan demo and neither are able to capture. From what I know the capture is supposed to happen during the call the **vkQueuePresentKHR** correct?

Are there any quirks to this? I have geometry rendering on screen but it doesn't appear to be inserting the capture boundaries...

**Comment by billhollings** [comment]
@MattGuerrette 

Are you able to get Frame Capture to work with the Demo apps that come with MoltenVK? Those definitely work as expected with frame capture.

When debugging your app...are you able to see and use the Xcode Frame Capture button?

What version of iOS or macOS are you testing against?

MoltenVK inserts capture boundaries at presentation submission...and on newer OS's (iOS 11 & macOS 10.13) at command buffer submissions as well.

**Comment by remoe** [comment]
Hi, i've the same issue. I try to find the part in the sample which enable Frame Capture in MoltenVK.

**Comment by MattGuerrette** [comment]
Hello, sorry for the long overdue update. I just attempted to run the API-Samples demo and capture a frame and I received the same capture boundary issue. However, I did try the cube demo and it worked. It seems like the issue could be related to static framework vs dynamic library?

**Comment by KenThomases** [comment]
Haven't looked into the situation with MoltenVK specifically, but I've found in the past that Xcode's GPU frame capture will only work if the executable loads the Metal framework when it's initially loaded (as a dependency library). Loading it later dynamically doesn't work. For the situation I was in, it was enough for me to simply link the program against Metal.framework even if nothing used it until a later library was dynamically loaded that also linked to Metal.framework.

**Comment by billhollings** [comment]
@MattGuerrette I am not having any trouble getting GPU frame capture to work with the Hologram demo (`libMoltenVK.a`)...the Cube (`libMoltenVK.dylib`)...or the Cube modified to use the framework (`MoltenVK.framework`).

Xcode 10.1
macOS 10.13.6

**Comment by billhollings** [comment]
Closing since this doesn't seem to be an issue with MoltenVK or the demos.

Please reopen if more info comes up.

**Comment by jamesd9** [comment]
@billhollings

I've run into this as well. I would reopen the issue but I don't seem to have the privileges to do so.

Note that my test case is using an iOS device with and OS version greater than kMinOSVersionMTLCaptureScope (the latter part in particular I suspect is important). It seems to be due to having multiple MVKQueues and the wrong one being the capture default.

To fix this I've changed initGPUCaptureScopes to take a bool that controls whether the queue should be made default:

```
void MVKQueue::initGPUCaptureScopes(bool makeDefault) {
	_submissionCaptureScope = new MVKGPUCaptureScope(this, "CommandBuffer-Submission");
	_presentationCaptureScope = new MVKGPUCaptureScope(this, "Surface-Presentation");
    if (makeDefault)
        _presentationCaptureScope->makeDefault();
    _presentationCaptureScope->beginScope();	// Allow Xcode to capture the first frame if desired.
}
```

Then the call to initGPUCaptureScopes from the MVKQueue constructor is modified as follows:

`initGPUCaptureScopes(queueFamily->getIndex() == 0);`

Obviously, this just hacks the first queue family to be the default, but this is likely to be better than the last queue family being the default and works for my case. I expect you'll be able to come up with a more elegant solution though. :)

**Comment by billhollings** [comment]
@jamesd9 

Thanks for pointing out the changes resulting from multiple queues.

This is now fixed in PR #463. MoltenVK config now identifies the queue family & queue to be used as the default presentation submission for GPU Capture scope. Please retest and confirm fix.

**Comment by jamesd9** [comment]
Thanks, that's fixed it. I don't have permission to close the bug though.
